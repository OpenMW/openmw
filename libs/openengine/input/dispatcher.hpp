#ifndef OENGINE_INPUT_DISPATCHER_H
#define OENGINE_INPUT_DISPATCHER_H

#include <cassert>
#include <iostream>

#include <mangle/input/event.hpp>

#include "dispatch_map.hpp"
#include "func_binder.hpp"

namespace OEngine {
namespace Input {

struct Dispatcher : Mangle::Input::Event
{
  DispatchMap map;
  FuncBinder funcs;

  /**
     Constructor. Takes the number of actions and passes it to
     FuncBinder.
  */
  Dispatcher(int actions) : funcs(actions) {}

  void bind(unsigned int action, int key)
  {
    assert(action < funcs.getSize());
    map.bind(key, action);
  }
  void unbind(unsigned int action, int key)
  {
    assert(action < funcs.getSize());
    map.unbind(key, action);
  }
  bool isBound(int key) const { return map.isBound(key); }

  /**
     Instigate an event. It is translated through the dispatch map and
     sent to the function bindings.
   */
  typedef DispatchMap::OutList _O;
  void event(Type type, int index, const void* p)
  {
    // No bindings, nothing happens
    if(!isBound(index))
      return;

    // Get the mapped actions and execute them
    const _O &list = map.getList(index);
    _O::const_iterator it;
    for(it = list.begin(); it != list.end(); it++)
    {
      //catch exceptions thrown in the input handlers so that pressing a key
      //doesn't cause OpenMw to crash
      try
      {
          funcs.call(*it, p);
      }
      catch(const std::exception& e)
      {
          std::cerr << "Exception in input handler: " << e.what() << std::endl;
      }
      catch(...)
      {
          std::cerr << "Unknown exception in input handler" << std::endl;
      }

    }
  }
};

// This helps us play nice with Mangle's EventPtr, but it should
// really be defined for all the classes in OEngine.
  typedef boost::shared_ptr<Dispatcher> DispatcherPtr;

}}
#endif
