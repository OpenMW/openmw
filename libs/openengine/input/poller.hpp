#ifndef OENGINE_INPUT_POLLER_H
#define OENGINE_INPUT_POLLER_H

#include "dispatch_map.hpp"
#include <mangle/input/driver.hpp>

namespace OEngine {
namespace Input {

/** The poller is used to check (poll) for keys rather than waiting
    for events. */
struct Poller
{
  DispatchMap map;
  Mangle::Input::Driver &input;

  Poller(Mangle::Input::Driver &drv)
    : input(drv) {}

  /** Bind or unbind a given action with a key. The action is the first
      parameter, the key is the second.
  */
  void bind(int in, int out) { map.bind(in, out); }
  void unbind(int in, int out) { map.unbind(in, out); }
  bool isBound(int in) const { return map.isBound(in); }

  /// Check whether a given action button is currently pressed.
  typedef DispatchMap::OutList _O;
  bool isDown(int index) const
  {
    // No bindings, no action
    if(!isBound(index))
      return false;

    // Get all the keys bound to this action, and check them.
    const _O &list = map.getList(index);
    _O::const_iterator it;
    for(it = list.begin(); it != list.end(); it++)
      // If there's any match, we're good to go.
      if(input.isDown(*it)) return true;

    return false;
  }
};
}}
#endif
