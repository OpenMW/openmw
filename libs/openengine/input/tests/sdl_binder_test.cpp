#include <iostream>
#include <mangle/input/servers/sdl_driver.hpp>
#include <SDL.h>
#include "../dispatcher.hpp"
#include "../poller.hpp"

using namespace std;
using namespace Mangle::Input;
using namespace OEngine::Input;

enum Actions
  {
    A_Quit,
    A_Left,
    A_Right,

    A_LAST
  };

bool quit=false;

void doExit(int,const void*)
{
  quit = true;
}

void goLeft(int,const void*)
{
  cout << "Going left\n";
}

int main(int argc, char** argv)
{
  SDL_Init(SDL_INIT_VIDEO);
  SDL_SetVideoMode(640, 480, 0, SDL_SWSURFACE);
  SDLDriver input;
  Dispatcher *disp = new Dispatcher(A_LAST);
  Poller poll(input);

  input.setEvent(EventPtr(disp));

  disp->funcs.bind(A_Quit, &doExit);
  disp->funcs.bind(A_Left, &goLeft);

  disp->bind(A_Quit, SDLK_q);
  disp->bind(A_Left, SDLK_a);
  disp->bind(A_Left, SDLK_LEFT);

  poll.bind(A_Right, SDLK_d);
  poll.bind(A_Right, SDLK_RIGHT);

  cout << "Hold the Q key to quit:\n";
  //input->setEvent(&mycb);
  while(!quit)
    {
      input.capture();
      if(poll.isDown(A_Right))
        cout << "We're going right!\n";
      SDL_Delay(20);

      if(argc == 1)
        {
          cout << "You are running in script mode, aborting. Run this test with a parameter (any at all) to test the input loop properly\n";
          break;
        }
    }
  cout << "\nBye bye!\n";

  SDL_Quit();
  return 0;
}
