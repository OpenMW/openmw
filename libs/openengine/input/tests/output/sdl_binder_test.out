Hold the Q key to quit:
You are running in script mode, aborting. Run this test with a parameter (any at all) to test the input loop properly

Bye bye!
